File: app/controllers/cotizador_controller.rb
# PR-10.b: endpoint JSON que alimenta el bloque "Valor a pagar" de
# /entrega_personal. Solo lectura, no persiste nada.
class CotizadorController < ApplicationController
  before_action :authorize_cotizador

  def show
    tipo_envio = TipoEnvio.find_by(id: params[:tipo_envio_id])
    return render(json: { ok: false, motivo: "sin_tipo_envio" }) if tipo_envio.nil?

    r = CotizadorFlete.call(
      tipo_envio: tipo_envio,
      cliente:    Cliente.find_by(id: params[:cliente_id]),
      proveedor:  Proveedor.find_by(id: params[:proveedor_id]),
      sucursal:   Sucursal.find_by(id: params[:sucursal_id]),
      peso: params[:peso], alto: params[:alto], largo: params[:largo], ancho: params[:ancho],
      # PR-C7.17: el envío completo cuando ya hay cajas cargadas. Llegan como
      # `cajas[N][campo]`, el mismo shape con el que viajan en el formulario.
      cajas: cajas_del_request
    )

    render json: {
      ok: true,
      tarifa_encontrada: r.tarifa_encontrada?,
      aplico_minimo: r.aplico_minimo,
      tasa: r.tasa.to_f,
      peso_cobrar: r.peso_facturado.to_f,
      vlbs: r.vlbs.to_f,
      # Nil cuando el envío tiene cajas en escalones distintos: ahí no hay "un"
      # precio por libra, y mandar el de una sola sería mentir.
      precio_libra: (par(r, r.precio_libra) if r.precio_libra),
      subtotal: par(r, r.subtotal),
      impuesto: par(r, r.impuesto),
      total: par(r, r.total)
    }
  end

  private

  CAMPOS_DE_CAJA = %w[peso alto largo ancho].freeze

  # Las cajas del envío, ordenadas por su número. Mismo criterio que
  # `MedidasPorCaja`: una caja sin ningún campo lleno no es una caja.
  def cajas_del_request
    crudo = params[:cajas]
    return [] if crudo.blank?

    crudo.to_unsafe_h
         .sort_by { |indice, _| indice.to_i }
         .filter_map { |_, valores| valores.slice(*CAMPOS_DE_CAJA).compact_blank.presence }
  end

  # Cada monto viaja en las dos monedas — "valor a pagar en dólares y
  # lempiras" (Yusef).
  def par(resultado, monto)
    { lps: monto.to_f, usd: resultado.en_usd(monto).to_f }
  end

  def authorize_cotizador
    require_role(:supervisor_miami, :digitador_miami, :supervisor_prefactura,
                 :supervisor_caja, :cajero)
  end
end

File: app/javascript/controllers/cotizador_controller.js
import { Controller } from "@hotwired/stimulus"

// PR-10.b: consulta el cobro estimado del flete mientras el operario llena el
// formulario. Yusef: "hay que agregarle para poder ver el valor a pagar...
// valor a pagar en dólares y lempiras".
//
// Solo display — el cobro definitivo lo calcula Pre-Factura. Este controller
// vive junto a `calc-volumetrico` en el mismo bloque; aquel hace el peso, este
// hace el dinero.
export default class extends Controller {
  static targets = ["panel", "precio", "subtotal", "isv", "total", "avisoMinimo", "nota"]
  static values = { url: String }

  connect() {
    // El cliente y el tipo de envío viven fuera de este bloque, así que se
    // escuchan a nivel de formulario en vez de por target.
    this._form = this.element.closest("form")
    this._onChange = () => this.cotizar()
    if (this._form) this._form.addEventListener("change", this._onChange)
    this.cotizar()
  }

  disconnect() {
    if (this._form) this._form.removeEventListener("change", this._onChange)
    clearTimeout(this._timeout)
  }

  cotizar() {
    clearTimeout(this._timeout)
    this._timeout = setTimeout(() => this._fetch(), 300)
  }

  async _fetch() {
    if (!this.hasUrlValue || !this._form) return

    const datos = new FormData(this._form)
    const params = new URLSearchParams()
    const campos = {
      tipo_envio_id: "paquete[tipo_envio_id]",
      cliente_id: "paquete[cliente_id]",
      proveedor_id: "paquete[proveedor_id]",
      sucursal_id: "paquete[sucursal_id]",
      peso: "paquete[peso]",
      alto: "paquete[alto]",
      largo: "paquete[largo]",
      ancho: "paquete[ancho]"
    }
    for (const [param, campo] of Object.entries(campos)) {
      const v = datos.get(campo)
      if (v) params.set(param, v)
    }

    // PR-C7.17: las cajas ya agregadas.
    //
    // Antes se cotizaba solo con los campos de captura, que `cajas-repetidor`
    // vacía al agregar — así que con dos cajas cargadas el cobro se desplomaba
    // al mínimo de servicio. Jorge lo vio: L.200 por un envío de 80 lb.
    //
    // Se mandan tal como viajan en el form (`paquete[cajas][N][campo]`) y el
    // servidor las totaliza. La suma NO se hace acá a propósito: la regla de
    // A9-03 —el mayor de cada caja, y después se suman— ya vive en Ruby, y una
    // tercera copia es la duplicación que PR-C7.11 vino a matar.
    let cajas = 0
    for (const [nombre, valor] of datos.entries()) {
      const m = nombre.match(/^paquete\[cajas\]\[(\d+)\]\[(\w+)\]$/)
      if (!m || !valor) continue
      params.append(`cajas[${m[1]}][${m[2]}]`, valor)
      cajas = Math.max(cajas, Number(m[1]))
    }
    // Con cajas cargadas manda el envío entero: la caja en curso todavía no es
    // una caja, y sumarla contaría lo que el operario aún está tecleando.
    if (cajas > 0) ["peso", "alto", "largo", "ancho"].forEach(c => params.delete(c))

    if (!params.get("tipo_envio_id")) return this._limpiar("Elegí el tipo de envío para ver el cobro.")

    try {
      const res = await fetch(`${this.urlValue}?${params}`, {
        headers: { Accept: "application/json" }
      })
      if (!res.ok) return this._limpiar("No se pudo calcular el cobro.")
      const d = await res.json()
      if (!d.ok) return this._limpiar("Elegí el tipo de envío para ver el cobro.")
      this._pintar(d)
    } catch (e) {
      console.warn("[cotizador] fallo la consulta:", e)
      this._limpiar("No se pudo calcular el cobro.")
    }
  }

  _pintar(d) {
    // Sin `precio_libra` el envío tiene cajas en escalones distintos: no hay un
    // precio por libra que mostrar. Se dice, en vez de dejar el de la consulta
    // anterior, que sería un número de otra cosa.
    if (d.precio_libra) {
      this._set(this.precioTarget, d.precio_libra)
    } else if (this.hasPrecioTarget) {
      this.precioTarget.textContent = "varía por caja"
    }
    this._set(this.subtotalTarget, d.subtotal)
    this._set(this.isvTarget, d.impuesto)
    this._set(this.totalTarget, d.total)

    if (this.hasAvisoMinimoTarget) {
      this.avisoMinimoTarget.classList.toggle("hidden", !d.aplico_minimo)
    }
    if (this.hasNotaTarget) {
      this.notaTarget.textContent = d.tarifa_encontrada
        ? `Tarifa del cliente · tasa L.${d.tasa}. El cobro definitivo se calcula en Pre-Factura.`
        : `Sin tarifa cargada para este servicio: se usa el precio base. Tasa L.${d.tasa}.`
    }
  }

  // "en dólares y lempiras" — los dos juntos, el lempira primero porque es la
  // moneda en la que se factura.
  _set(target, par) {
    if (!target || !par) return
    target.textContent = `L. ${this._fmt(par.lps)} · $${this._fmt(par.usd)}`
  }

  _limpiar(mensaje) {
    for (const t of [this.precioTarget, this.subtotalTarget, this.isvTarget, this.totalTarget]) {
      if (t) t.textContent = "—"
    }
    if (this.hasAvisoMinimoTarget) this.avisoMinimoTarget.classList.add("hidden")
    if (this.hasNotaTarget && mensaje) this.notaTarget.textContent = mensaje
  }

  _fmt(n) {
    return Number(n || 0).toLocaleString("en-US", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    })
  }
}
